src/ralph/lib/polymorphic/models.py

```diff
@@ -0,0 +1,118 @@
+# -*- coding: utf-8 -*-
+"""
+Polymorphic using inherited models in Django.
+Displays information in queryset list about the model that inherits
+the polymorphic model
+
+Example:
+    >>>  Model.polymorphic_objects.all()
+    [
+        <Model1: model1: test>,
+        <Model2: model2: test>,
+        <Model3: model3L test>
+    ]
+"""
+from itertools import groupby
+
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+from django.db.models.fields.related import OneToOneField
+
+
+class PolymorphicQuerySet(models.QuerySet):
+
+    def iterator(self):
+        """
+        Override iterator:
+            - Iterate for all objects and collected ID
+            - For each ContentType generates additional queryset
+            - Returns iterator with different models
+        """
+        result = []
+        content_type_model_map = {}
+        for obj in super().iterator():
+            content_type_model_map[
+                obj.content_type_id
+            ] = obj.content_type.model_class()
+            result.append((
+                obj.content_type_id, obj.pk)
+            )
+        result = groupby(result, lambda x: x[0])
+
+        result_query = []
+        for k, v in result:
+            model = content_type_model_map[k]
+            polymorphic_models = getattr(model, '_polymorphic_models', [])
+            if polymorphic_models and model not in polymorphic_models:
+                result_query.extend(
+                    model.objects.filter(pk__in=[i[1] for i in v])
+                )
+
+        for obj in result_query:
+            yield obj
+
+
+class PolymorphicBase(models.base.ModelBase):
+    def __new__(cls, name, bases, attrs):
+        full_mro = set(
+            tuple([mro for b in bases for mro in b.__mro__]) + bases
+        )
+        base_polymorphic = set(
+            [b for b in full_mro if issubclass(b, Polymorphic)]
+        )
+        # Set is_polymorphic flag for classes that use polymorphic
+        for polymorphic_class in base_polymorphic:
+            polymorphic_class.is_polymorphic = True
+        attrs['_polymorphic_models'] = base_polymorphic
+        return super().__new__(cls, name, bases, attrs)
+
+
+class Polymorphic(models.Model):
+
+    """
+    Polymorphic model.
+
+    Added content type field to model
+
+    Example:
+        >>> class Test(Polymorphic, models.Model, metaclass=PolymorphicBase):
+                pass
+    """
+
+    content_type = models.ForeignKey(ContentType, blank=True, null=True)
+
+    polymorphic_objects = PolymorphicQuerySet.as_manager()
+    objects = models.Manager()
+
+    class Meta:
+        abstract = True
+
+    def save(self, *args, **kwargs):
+        """
+        Save object content type
+        """
+        if not self.content_type:
+            self.content_type = ContentType.objects.get_for_model(self)
+        super().save(*args, **kwargs)
+
+    @classmethod
+    def get_descendants_models(cls):
```

Comment:
in my opinion we should do this only once per model and it should be done in `PolymorphicBase` (when you're iterating over `base_polymorphic` you could store additional variable in each of this classes list, to which you will append currently created class), ex:

```
attrs['_polymorphic_descendats'] = []
new_class = super().__new__(cls, name, bases, attrs)  # notice calling super before loop
for polymorphic_class in base_polymorphic:
    polymorphic_class._is_polymorphic = True  # notice _ here at the beginning
    polymorphic_class._polymorphic_descendats.append(new_class)
return new_class
```